Show Peer Responses on page load if q answered

caesar2164 opened the pull request:
- on page load make ajax call with student answer
- if student answer is valid display peer responses
- also defines a function to check for response validity and displaying peer responses that gets used on both page load and submit click

@stvstnfrd this was remarkably easy...

Commit:
Show Peer Responses on page load if q answered

- Hide responses if feature turned off
- Hide responses on page load if NOT answered
- Display responses is student answer valid

stvstnfrd commented on freetextresponse/public/view.js:
```diff
@@ -139,4 +144,24 @@ function FreeTextResponseView(runtime, element) {
+                'student_answer': $element.find('.student_answer').val()
+            }),
+            success: function xblockLoadOnSuccess(response) {
+                display_responses_if_answered(response);
```
Instead of having two separate AJAX calls, can we refactor the two calls into a single helper?

Also, why are we waiting to make an AJAX request _after_ page load;
this means you're going to get a jump/flash of content when the ajax call finally goes through.

Since we already know at initial page load whether or not we should already display the results,
we should just include them directly in the original page load/response.

caesar2164 replied:
I'm not sure if refactoring those two would net us any benefit right now. There are a bunch of differences both in data being sent, as well as in success reactions. (I already factored out the overlapping stuff into a helper)

Unfortunately we CAN'T do the AJAX request any earlier, XBlocks get loaded AFTER page load. Since I need to query the XBlock to get its answer, as well as peer responses, I HAVE to wait for the XBlock to be loaded like I do...

stvstnfrd replied:
> Unfortunately we CAN'T do the AJAX request any earlier

I think you may have missed my point:

> we should just include them directly in the original page load/response.

I'm suggesting that you _not_ use AJAX _at_ _all_ for initializing the XBlock.

caesar2164 replied:
The problem is that `response.other_responses` is only available in an AJAX response... so we CAN'T know if/what we should be displaying without an AJAX request.

caesar2164 replied:
ok, I guess we could, preload the HTML with answers, but I'm actually not sure if we should?

stvstnfrd commented:
@caesar2164 ping.

caesar2164 commented:
nvmd, brain fart re AJAX vs page load...

caesar2164 commented on freetextresponse/public/view.js:
```diff
@@ -98,6 +93,16 @@ function FreeTextResponseView(runtime, element) {
         return html;
     }
 
+    function display_responses_if_answered(response) {
```
Should I remove this helper function?

stvstnfrd replied:
I wouldn't.

Honestly, I tend to find that anything more than 5 (ish) lines can generally be pulled out into a dedicated function.

Here, this is a great level of abstraction; it's a helper function that makes it abundantly clear what's going on.
Otherwise, you end up needing to overly comment each stanza, something like:

Note: lines with `###` are placeholders for code, while single `#` are actually comments that would be added to the code

```python
def main():
    # initialize a bunch of stuff
    pass  ### ... lots of lines
    
    # fetch data
    pass  ### make ajax calls and whatever

    # parse data
    pass  ### several lines looping over data and processing it

    # display data
    pass  ### all the stuff in `display_responses_if_answered`
```

Now compare that to this version w/ helpers:

```python
def main():
    self.initialize_state()
    data = fetch_data()
    data = parse_data(data)
    display_responses_if_answered(data)
```

Well-named helper functions give us automatic documentation :)

stvstnfrd replied:
Also, see note about JS using camelCase instead of snake_case.

caesar2164 replied:
Resolved

caesar2164 commented:
ok @stvstnfrd ready for round 2

stvstnfrd commented on freetextresponse/freetextresponse.py:
```diff
@@ -279,6 +279,17 @@ def student_view(self, context={}):
+            other_responses = self.get_other_answers(student_id)
+        else:
+            other_responses = []
+        if other_responses:
```
Since this `if` check is purely presentational, shouldn't we move it to the template?

Instead of the backend telling the frontend whether or not to display an element,
we should instead delegate that decision to the template.

So then the template can rationalize,
"if I don't have other responses, I shouldn't try to display them".

stvstnfrd commented on freetextresponse/freetextresponse.py:
```diff
@@ -279,6 +279,17 @@ def student_view(self, context={}):
+        student_id = self.get_student_id()
+        display_other_responses = self.display_other_student_responses
+        student_answer = self.student_answer
+        if student_answer and display_other_responses:
```
Kind of a nit, but I find it to be a little safer to explicitly declare a default value, instead of relying on the `else` clause. It also reads a bit more cleanly like:

```python
other_responses = []
if student_answer and display_other_responses:
    other_responses = self.get_other_answers(student_id)
```

stvstnfrd commented on freetextresponse/public/view.js:
```diff
@@ -98,6 +93,16 @@ function FreeTextResponseView(runtime, element) {
+            if (responseHTML) {
+                responseList.html(responseHTML);
+            }
+            $element.find('.responses-box').removeClass('hidden');
```
Should this instead be inside the `if` block?
That way it only tries to remove the class if it actually pulled back HTML?

Or actually, what happens if the AJAX lookup here (from `get_student_responses_html`) doesn't find any items. Should `get_student_responses_html` instead return the "No messages to display" instead of an empty string? In that case, we can remove the `if` check completely and just update the inner html regardless.

caesar2164 replied:
By default `responseList` is loaded with `<li class="no-response">No responses to show at this time</li>` if there isn't a student response. So if the AJAX call comes back empty it'll just show that default value...

stvstnfrd replied:
I think that relies too much on assumptions of the underlying implementation.

If we don't have any `responseHTML` and we aren't updating the content of `responseList`,
then we shouldn't be modifying the responses box.

It's possible, however unexpected, that you could start with responses, and later get back nothing.
- a response could be deleted
- an error could occur that causes the lookup to fail and instead returns an empty list

We can't make this assumption.

An easy way to do this is to just bail immediately if there isn't anything to update:

```javascript
if (!responseHtml) {
    return;
}
responseList.html(responseHtml);
$element.find('.responses-box').removeClass('hidden');
```

But still, as pointed out, this doesn't address what happens if we _do_ have responses at one point, but then later we don't.
The easy way to handle this is to just have our `get_student_responses_html` helper return either the response html or the "no-response" html.

stvstnfrd replied:
We can discuss more f2f, if needed.

stvstnfrd commented on freetextresponse/templates/freetextresponse_view.html:
```diff
@@ -30,11 +30,15 @@ <h2 class="problem-header">{{ display_name }}</h2>
     <div class="capa_alert user_alert">{{ user_alert }}</div>
     {% if display_other_responses %}
-        <div class="responses-box">
+        <div class="responses-box {{ hide_responses_class }}">
```
See note above about moving display logic into here.

caesar2164 commented:
@stvstnfrd I think I've addressed everything. (I also rebased to resolve a conflict from the previous merged bugfix)

stvstnfrd commented on freetextresponse/templates/freetextresponse_view.html:
```diff
@@ -30,16 +30,24 @@ <h2 class="problem-header">{{ display_name }}</h2>
-            </ul>
-        </div>
+        {% if student_answer %}
+            <div class="responses-box">
```
I think this got over-engineered.
What you had here (class name in a variable) previously was fine.
In fact, it's what we want, to specify visibility with a variable.

I was just pointing out that it should be the template itself which assigns the variable,
not the backend.

stvstnfrd commented on freetextresponse/public/view.js:
```diff
@@ -102,6 +97,16 @@ function FreeTextResponseView(runtime, element) {
 
+    function display_responses_if_answered(response) {
+        if (!response.user_alert && response.display_other_responses) {
+            var responseHTML = get_student_responses_html(response.other_responses);
```
And while we're in here, let's explicitly hoist the variable to the top of the function and update our naming to match JS convention (camelCase instead of snake_case).

`responseHtml` and `getStudentResponsesHtml`.

stvstnfrd commented on freetextresponse/templates/freetextresponse_view.html:
```diff
@@ -30,16 +30,24 @@ <h2 class="problem-header">{{ display_name }}</h2>
+                    {% for response in other_responses %}
+                        <li class="other-student-responses">{{ response.answer }}</li>
+                    {% empty %}
+                        <li class="no-response">No responses to show at this time</li>
```
Does i18n happen automatically or does it need specified?

caesar2164 commented:
@stvstnfrd ready for yet another round of review. (Diff down to +19!)

stvstnfrd commented on freetextresponse/freetextresponse.py:
```diff
@@ -660,6 +660,12 @@ def get_other_answers(self, student_id):
+        student_hasnt_answered_properly = self._determine_credit() == Credit.zero
+        if student_hasnt_answered_properly or shouldnt_show_other_responses:
+            return []
+
```
Nit: I'd remove the empty line..

stvstnfrd commented on freetextresponse/templates/freetextresponse_view.html:
```diff
@@ -30,15 +30,19 @@ <h2 class="problem-header">{{ display_name }}</h2>
     <div class="capa_alert user_alert">{{ user_alert }}</div>
     {% if display_other_responses %}
-        <div class="responses-box">
+        <div class="responses-box {% if not student_answer %}hidden{% endif %}">
```
Nice!

stvstnfrd commented on freetextresponse/freetextresponse.py:
```diff
@@ -294,7 +294,7 @@ def student_view(self, context={}):
                 'word_count_message': self._get_word_count_message(),
                 'display_other_responses': self.display_other_student_responses,
-                'other_responses': [],
+                'other_responses': self.get_other_answers(self.get_student_id()),
```
Since we're never re-using a memoized version of this, just move the `get_student_id` lookup into the method itself.

stvstnfrd commented on freetextresponse/public/view.js:
```diff
@@ -102,6 +96,18 @@ function FreeTextResponseView(runtime, element) {
+            return;
+        }
+        var responseHTML = get_student_responses_html(response.other_responses);
+        if (responseHTML) {
```
I have a question about this, I'll just discuss f2f..

stvstnfrd commented:
@caesar2164 What's the status of this?

caesar2164 commented:
waiting for your review @stvstnfrd

stvstnfrd commented on freetextresponse/public/view.js:
```diff
@@ -99,9 +93,20 @@ function FreeTextResponseView(runtime, element) {
         responses.forEach(function(item) {
             html += '<li class="other-student-responses">' + item.answer + '</li>';
         });
+        html = html || '<li class="no-response">No responses to show at this time</li>';
```
What about i18n?

caesar2164 replied:
unfortunately that doesn't work in xBlocks yet: https://openedx.atlassian.net/browse/PLAT-1141

stvstnfrd replied:
@caesar2164 Then just stuff it into the HTML template as a hidden element and grab it to use as a template here.

stvstnfrd commented on freetextresponse/public/view.js:
```diff
@@ -99,9 +93,20 @@ function FreeTextResponseView(runtime, element) {
         return html;
     }
 
+    function display_responses_if_answered(response) {
```
Javascript naming convention is camelCase not snake_case.

stvstnfrd commented on freetextresponse/public/view.js:
```diff
@@ -99,9 +93,20 @@ function FreeTextResponseView(runtime, element) {
+            $element.find('.responses-box').addClass('hidden');
+            return;
+        }
+        var responseHTML = get_student_responses_html(response.other_responses);
```
Update style convention for JS: `responseHtml`...

stvstnfrd reviewed (changes requested):
Some small things, mostly style, otherwise looks good.

Can you do a final demo for Jen, Monica, and Kailey?

stvstnfrd commented on freetextresponse/mixins.py:
```diff
@@ -33,7 +33,7 @@ def is_past_due(self):
 
 
-class MissingDataFetcherMixin(object):
+class MissingDataFetcherMixin(object):  # pylint: disable=too-few-public-methods
```
You should be able to also disable the long line warning here.

stvstnfrd commented:
@caesar2164 I resolved most of the convos, just a few notes left about i18n and pep8.

stvstnfrd commented on freetextresponse/mixins.py:
```diff
@@ -33,7 +33,7 @@ def is_past_due(self):
 
 
-class MissingDataFetcherMixin(object):
+class MissingDataFetcher(object):  # pylint: disable=too-few-public-methods
```
Eh, I don't know how I feel about renaming a sensible variable just because a pylint comment makes a line too long..

stvstnfrd replied:
What if you move the disable to above the `class` declaration instead of beside it?

It's a little more broad, but really, it could apply to any/all classes in this file; that's fine.

stvstnfrd replied:
And better yet, I think you can re-enable the flag again right after this class:

`# pylint: enable=too-foo-public-methods`

stvstnfrd commented on freetextresponse/templates/freetextresponse_view.html:
```diff
@@ -30,15 +30,19 @@ <h2 class="problem-header">{{ display_name }}</h2>
+                {% for response in other_responses %}
+                    <li class="other-student-responses">{{ response.answer }}</li>
+                {% empty %}
+                    <li class="no-response">No responses to show at this time</li>
```
Maybe always write this item out to the template, `display: none;` by default, like before,
then display it when when `len(other_responses) == 0`.

stvstnfrd replied:
Sorry, you'd want to write it out so that you can grab it elsewhere in your javascript..

stvstnfrd commented on freetextresponse/freetextresponse.py:
```diff
@@ -652,14 +654,21 @@ def store_student_response(self):
         # Want to store extra response so student can still see
         # MAX_RESPONSES answers if their answer is in the pool.
-        self.displayable_answers = self.displayable_answers[-(MAX_RESPONSES+1):]
+        response_key = -(MAX_RESPONSES+1)
```
This is a number, right?
Should it be `response_index`?

stvstnfrd commented on freetextresponse/freetextresponse.py:
```diff
@@ -652,14 +654,21 @@ def store_student_response(self):
+        student_id = self.get_student_id()
+        display_other_responses = self.display_other_student_responses
+        shouldnt_show_other_responses = not display_other_responses
+        student_answer_not_proper = self._determine_credit() == Credit.zero
```
`student_answer_incorrect`?

"Proper" seems ambiguous...

Commit:
Allow for translation of text in HTML template

Commit:
Fix pep8 and pylint errors in modified files

- define new variables to shorten line lengths
- remove unused import
- indent arguments
- remove empty lines
- disable too-few-public-methods where pertinent
